src/app/shared/services/media-state.service.ts

```diff
@@ -1,33 +1,81 @@
+  ) {
+    this.logger = logger.tag('MediaStateService')
+    lastValueFrom(this.mediaService.getAllMedia())
+      .then(mediaAssets => mediaAssets.forEach(media => this.updateMedia(media)))
+      .catch(error => this.logger.data(error).error(`Failed while fetching all media from server`))
+    this.subscribeToEvents()
+  }
+
+  private updateMedia(media: Media): void {
+    const mediaSubject: BehaviorSubject<Media | undefined> | undefined = this.mediaSubjects.get(media.sourceName)
+    if (mediaSubject) {
+      mediaSubject.next(media)
+      return
+    }
+    this.mediaSubjects.set(this.getMediaSubjectKey(media.sourceName), new BehaviorSubject<Media | undefined>(media))
+  }
+
+  private getMediaSubjectKey(sourceName: string): string {
+    return sourceName.toUpperCase()
+  }
 
-  public async subscribeToMedia(id: string): Promise<Observable<Media>> {
-    const mediaSubject: BehaviorSubject<Media> = await this.createMediaSubject(id)
-    return mediaSubject.asObservable()
+  private subscribeToEvents(): void {
+    this.subscriptions.push(this.mediaEventObserver.subscribeToMediaCreated((event: MediaCreatedEvent) => this.createMediaFromEvent(event)))
+    this.subscriptions.push(this.mediaEventObserver.subscribeToMediaDeleted((event: MediaDeletedEvent) => this.deleteMediaFromEvent(event)))
   }
 
-  private async createMediaSubject(id: string): Promise<BehaviorSubject<Media>> {
-    const mediaSubject: BehaviorSubject<Media> | undefined = this.mediaSubject.get(id)
+  private createMediaFromEvent(event: MediaCreatedEvent): void {
+    const mediaSubject: BehaviorSubject<Media | undefined> | undefined = this.mediaSubjects.get(event.media.sourceName)
     if (mediaSubject) {
-      return mediaSubject
+      mediaSubject.next(event.media)
+      return
     }
-    return await this.getCleanMediaSubject(id)
+    this.mediaSubjects.set(event.media.sourceName, new BehaviorSubject<Media | undefined>(event.media))
   }
 
-  private async getCleanMediaSubject(id: string): Promise<BehaviorSubject<Media>> {
-    const media: Media = await this.fetchMedia(id)
-    return new BehaviorSubject<Media>(media)
+  private deleteMediaFromEvent(event: MediaDeletedEvent): void {
+    this.mediaSubjects.forEach(subject => {
+      if (subject.value?.id === event.mediaId) {
+        subject.next(undefined)
+      }
+    })
+  }
+
+  public subscribeToMedia(id: string): Observable<Media | undefined> {
+    return this.createMediaSubject(id).asObservable()
+  }
+
+  private createMediaSubject(sourceName: string): BehaviorSubject<Media | undefined> {
+    const mediaSubjectKey: string = this.getMediaSubjectKey(sourceName)
+    const mediaSubject: BehaviorSubject<Media | undefined> | undefined = this.mediaSubjects.get(mediaSubjectKey)
+    if (mediaSubject) {
+      return mediaSubject
+    }
+    const subject: BehaviorSubject<Media | undefined> = new BehaviorSubject<Media | undefined>(undefined)
+    this.mediaSubjects.set(mediaSubjectKey, subject)
```

Comment:
This is the third time in this file where you set `this.mediaSubjects.set(..., ...)` i.e. this the third time where you create a new BehaviourSubject.

Why not have something like:
```
private getMediaSubject(sourceName: string): BehaviourSubject<Media | undefined> {
  if (!this.mediaSubjects.has(this.getMediaSubjectKey(sourceName)) {
    this.mediaSubjects.set(this.getMediaSubjectKey(sourceName), new BehaviourSubject<Media | undefined>(undefined)
  }
  return this.mediaSubjects(this.getMediaSubjectKey(sourceName)!
}
```

Then your `updateMedia` method becomes:
```
private updateMedia(media: Media): void {
  const mediaSubject = this.getMediaSubject(media.sourceName)
  mediaSubject.next(media)
}
```

And your `createMediaFromEvent` method becomes: 
```
private createMediaFromEvent(event: MediaCreatedEvent): void {
  const mediaSubject = this.getMediaSubject(event.media.sourceName)
  mediaSubject.next(event.media)
}
```

And `subscribteToMedia` becomes:
```
public subscribeToMedia(sourceName: string): Observable<Media | undefined> {
  return this.createMediaSubject(sourceName).asObservable()
}
```

Reply:
Notice how identical `updateMedia` and `createMediaFromEvent` is?
This would allow you to delete `createMediaFromEvent` and just use `updateMedia`:
```
this.subscriptions.push(this.mediaEventObserver.subscribeToMediaCreated((event: MediaCreatedEvent) => this.updateMedia(event.media)))
```

Further emphaszing why we shouldn't name stuff like `...FromEvent` because it is based on _where_/_how_ we want to use our methods instead of what they actually _do_.

Reply:
I made this refactor, a great suggestion and i must admit i should have thought more about the `FromEvent` naming  also!

Reply:
I agree with the overall assessment, except with the statement that we shouldn't use `...FromEvent`. It does declare a premise for how it is used, but not _where_ it is used. The given alternative is to convert the method to a anonymous function. Personally I prefer `...FromEvent` over `on<Event>` syntax, since it describes it intent and what the responsibility is: Namely, unwrapping the event the event for the logic.

Reply:
It does declare a premise for _where_ it is used, but it does it by saying where it _can't_ be used.

The current implementation, after Rasmus refactor, has the method `updateMedia` which is used in two places. If you renamed it back to `updateMediaFromEvent` you can now only use it in _one_ of those places, because only one those places are from an event. So now you need a second method to do the same thing as `updateMediaFromEvent`.

But the `fromEvent` part still refers to something _outside_ of the scope of the method. The method does not and should not care about where it is called from as long as its contract is fullfilled (in this case, it gets a Media, doesn't matter where from). So the `fromEvent` is redundant and it limits the logical reuse of the method.

Reply:
I followed Lindveds logic here because what he mentions about `updateMedia`, but also because the other method used for events is deleting a subject and that could easily become applicable in other cases where we would have the same issue. I understand wanting to signify the fact origin of something, but then we would have to accept that we need two methods that do really just do the same thing.

Reply:
`FromEvent` does not refer where it is used, but defines what type of data it works with. That will naturally limit the places where it can be used. We also use this pattern e.g. with `createPartFromAction` in Sofie server. `FromAction` specifies what data the method works on, in the same manner as `FromEvent` does it.
If it should be used here, then I agree, that the duplication in `updateMedia` and `updateMediaFromEvent` can be combined, but in that case it would become:
```typescript
private updateMediaFromEvent(event: MediaCreatedEvent): void {
  this.updateMedia(event.media)
}

private updateMedia(media: Media): void {
  const mediaSubject = this.getMediaSubject(media.sourceName)
  mediaSubject.next(media)
}
```

Reply:
> That will naturally limit the places where it can be used.

I.e. referencing _where_ it can and can't be used, not _what_ the method does...

Your `updateMediaFromEvent` method is redundant. It's just a parsethrough method that take something from the parameter and parses it along.
You want to write:
```
this.subscriptions.push(this.mediaEventObserver.subscribeToMediaCreated((event: MediaCreatedEvent) => this.updateMediaFromEvent(event)))
```
When you could just write:
```
this.subscriptions.push(this.mediaEventObserver.subscribeToMediaCreated((event: MediaCreatedEvent) => this.updateMedia(event.media)))
```